src/utils/presence.js

```diff
@@ -28,31 +29,52 @@ const OFFLINE_THRESHOLD_SECS = 140;
  *   https://zulip.readthedocs.io/en/latest/subsystems/presence.html
  *
  * This logic should match `status_from_timestamp` in the web app's
- * `static/js/presence.js`.
+ * `static/js/presence.js` at 1ae07b93d^.
  */
-export const getAggregatedPresence = (presence: UserPresence): ClientPresence =>
-  /* Out of the ClientPresence objects found in `presence`, we consider only
-   * those with a timestamp newer than OFFLINE_THRESHOLD_SECS; then of
-   * those, return the one that has the greatest PresenceStatus, where
-   * `active` > `idle` > `offline`.
+export const getAggregatedPresence = (presence: UserPresence): ClientPresence => {
+  /* Gives an artificial ClientPresence object where
+   * - `timestamp` is the latest `timestamp` of all ClientPresence
+   *   objects in the input, and
+   * - `status` is the greatest `status` of all *recent* ClientPresence
+   *   objects, where `active` > `idle` > `offline`. If there are no
+   *   recent ClientPresence objects (i.e., they are all at least as
+   *   old as OFFLINE_THRESHOLD_SECS), `status` is 'offline'. If there
+   *   are several ClientPresence objects with the greatest
+   *   PresenceStatus, an arbitrary one is chosen.
+   * - `client` is the `client` of the ClientPresence object from the
+   *   calculation of `status` (see previous), or empty string if
+   *   there wasn't one.
    *
-   * If there are several ClientPresence objects with the greatest
-   * PresenceStatus, an arbitrary one is chosen.
+   * It may not be entirely correct, but the `timestamp`'s
+   * ClientPresence object may be a different one than what `status`
+   * and `client` were drawn from.
    */
-  Object.keys(presence)
-    .filter((client: string) => client !== 'aggregated')
-    .reduce(
-      (aggregated, client: string) => {
-        const { status, timestamp } = presence[client];
-        if (Date.now() / 1000 - timestamp < OFFLINE_THRESHOLD_SECS) {
-          if (presenceStatusGeq(status, aggregated.status)) {
-            return { client, status, timestamp };
-          }
-        }
-        return aggregated;
-      },
-      { client: '', status: 'offline', timestamp: 0 },
-    );
+
+  let status = 'offline';
+  let client = '';
+  let timestamp = 0;
+
+  const dateNow = Date.now();
```

Comment:
small thing in commit message:

    This is a fresh optimization that wasn't done before this series of
    commits, so it gets its own commit.

This isn't really an optimization -- it's about correctness and consistent behavior. In this function in particular, it avoids the possibility of treating one presence report as too old and another as not too old, when the latter is actually older than the former. That's... harmless in this case, I think, but it's best not to have to worry about that class of things in general.

Reply:
Absolutely!